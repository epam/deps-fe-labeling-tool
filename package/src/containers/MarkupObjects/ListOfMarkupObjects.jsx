import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { toggleExpandedListKey } from '@/actions/ui'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape, Label } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { expandedListKeysSelector } from '@/selectors/ui'
import { replaceAll } from '@/utils/string'
import { ListContentWrapper, ListName, Collapse } from './ListOfMarkupObjects.styles'

const RESOURCE_LIST_NAME = 'List of {0}s'

const getUniqueSets = (markupObjects) => {
  const uniqueSets = []

  markupObjects.forEach((currentLabel) => {
    const hasDuplicate = uniqueSets.some((set) => set.includes(currentLabel))
    if (hasDuplicate) {
      return
    }

    const sameLabels = Label.getSameLabels(currentLabel, markupObjects)
    uniqueSets.push(sameLabels)
  })

  return uniqueSets
}

const ListOfMarkupObjects = ({
  markupObjects,
  field,
  renderMapper,
  expandedListKeys,
  toggleExpandedListKey
}) => {
  const renderer = renderMapper[field.fieldMeta.baseType]

  const onChangeHandler = useCallback(() => {
    toggleExpandedListKey(field.code)
  }, [field.code, toggleExpandedListKey])

  const setsOfLabels = getUniqueSets(markupObjects)

  return (
    <Collapse
      activeKey={expandedListKeys.find((key) => key === field.code)}
      onChange={onChangeHandler}
      header={
        <ListName>
          {replaceAll(RESOURCE_LIST_NAME, field.name)}
        </ListName>
      }
      collapseId={field.code}
    >
      <ListContentWrapper>
        {
          field.fieldMeta.baseType === FieldType.PAIR
            ? renderer(markupObjects, field)
            : setsOfLabels.map((labels, i) => renderer(labels, field, i))
        }
      </ListContentWrapper>
    </Collapse>
  )
}

ListOfMarkupObjects.propTypes = {
  markupObjects: PropTypes.arrayOf(
    PropTypes.oneOfType([
      labelShape,
      tableShape
    ])
  ),
  field: fieldShape,
  toggleExpandedListKey: PropTypes.func.isRequired,
  renderMapper: PropTypes.shape({
    [PropTypes.string]: PropTypes.func
  }).isRequired,
  expandedListKeys: PropTypes.arrayOf(PropTypes.string)
}

const mapStateToProps = (state) => ({
  expandedListKeys: expandedListKeysSelector(state)
})

const mapDispatchToProps = {
  toggleExpandedListKey
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ListOfMarkupObjects)

export {
  ConnectedComponent as ListOfMarkupObjects
}
