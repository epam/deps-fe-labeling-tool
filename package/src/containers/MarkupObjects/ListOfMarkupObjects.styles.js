
import styled from 'styled-components'
import { Collapse } from '@/components/Collapse'
import { COLORS } from '@/theme/theme.default'

const NameSpan = styled.span`
  font-size: 1.2rem;
  line-height: 1.6rem;
  color: ${COLORS.PRIMARY_4};
`

const ListName = styled(NameSpan)`
  align-items: center;
  font-size: 1.4rem;
  line-height: 2.2rem;
  font-weight: 600;
`

const ListContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.8rem;
  margin-left: 0.8rem;
`

const MarkupCollapse = styled(Collapse)`
  align-items: center;
`

export {
  ListName,
  ListContentWrapper,
  MarkupCollapse as Collapse
}
