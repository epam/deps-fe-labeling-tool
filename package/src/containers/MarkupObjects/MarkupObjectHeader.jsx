import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { clearSelection } from '@/actions/markup'
import { openPage } from '@/actions/pagination'
import { changeTool } from '@/actions/tools'
import { setActiveSidebar } from '@/actions/ui'
import { TextHighlighter } from '@/components/TextHighlighter'
import { SidebarContent } from '@/enums/SidebarContent'
import { Tool } from '@/enums/Tool'
import { labelShape } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { pageSelectedLabelsSelector, pageSelectedTablesSelector } from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import { selectedToolSelector } from '@/selectors/tools'
import { markupObjectsFilterSelector } from '@/selectors/ui'
import { childrenShape } from '@/utils/propTypes'
import { replaceAll } from '@/utils/string'
import { TextWrapper, TitleWithIconWrapper, PageSpan, NameSpan } from './MarkupObjectHeader.styles'

const RESOURCE_PAGE_1 = 'page {0}'

const MarkupObjectHeader = ({
  badge,
  collapseIcon,
  icon,
  name,
  pagesTitle,
  unassigned,
  page,
  markupObjectFilter,
  onHeaderClick,
  currentPage,
  clearSelection,
  openPage,
  changeTool,
  selectedTool,
  selectedLabels,
  selectedTables,
  setActiveSidebar
}) => {
  const onClick = () => {
    if (selectedTool !== Tool.POINTER) {
      changeTool(Tool.POINTER)
    }

    if (selectedLabels.length || selectedTables.length) {
      clearSelection(currentPage)
    }

    setActiveSidebar(SidebarContent.MARKUP)

    if (currentPage !== page && page !== undefined) {
      openPage(page)
    }

    onHeaderClick?.()
  }

  return (
    <TextWrapper
      $unassigned={unassigned}
      onClick={onClick}
    >
      {badge}
      {collapseIcon}
      <TitleWithIconWrapper>
        {icon}
        <NameSpan>
          <TextHighlighter
            searchTerm={markupObjectFilter}
            text={name}
          />
        </NameSpan>
      </TitleWithIconWrapper>
      {pagesTitle && <PageSpan>{pagesTitle}</PageSpan>}
      {page && <PageSpan>{replaceAll(RESOURCE_PAGE_1, page)}</PageSpan>}
    </TextWrapper>
  )
}

MarkupObjectHeader.propTypes = {
  badge: childrenShape,
  collapseIcon: childrenShape,
  icon: childrenShape,
  pagesTitle: PropTypes.string,
  setActiveSidebar: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  page: PropTypes.number,
  unassigned: PropTypes.bool,
  markupObjectFilter: PropTypes.string,
  onHeaderClick: PropTypes.func,
  currentPage: PropTypes.number.isRequired,
  openPage: PropTypes.func.isRequired,
  clearSelection: PropTypes.func.isRequired,
  changeTool: PropTypes.func.isRequired,
  selectedTool: PropTypes.oneOf(
    Object.values(Tool)
  ).isRequired,
  selectedLabels: PropTypes.arrayOf(labelShape).isRequired,
  selectedTables: PropTypes.arrayOf(tableShape).isRequired
}

const mapStateToProps = (state) => ({
  markupObjectFilter: markupObjectsFilterSelector(state),
  currentPage: currentPageSelector(state),
  selectedTool: selectedToolSelector(state),
  selectedLabels: pageSelectedLabelsSelector(state),
  selectedTables: pageSelectedTablesSelector(state)
})

const mapDispatchToProps = {
  openPage,
  clearSelection,
  changeTool,
  setActiveSidebar
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(MarkupObjectHeader)

export {
  ConnectedComponent as MarkupObjectHeader
}
