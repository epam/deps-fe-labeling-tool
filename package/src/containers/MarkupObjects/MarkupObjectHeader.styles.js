
import styled, { css } from 'styled-components'
import { COLORS } from '@/theme/theme.default'

const TextWrapper = styled.div`
  display: flex;
  flex: 1 1 auto;
  align-items: center;
  position: relative;
  padding: 0.8rem 1.6rem;
  border: 1px solid ${COLORS.GRAYSCALE_1};
  border-radius: 0.4rem;
  ${(props) => props.$unassigned && css`
      background: ${COLORS.GRAYSCALE_1};
    `}
`

const TitleWithIconWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 1.3rem;
`

const NameSpan = styled.span`
  font-size: 1.4rem;
  line-height: 2.2rem;
  font-weight: 600;
  color: ${COLORS.PRIMARY_4};
`

const PageSpan = styled.span`
  margin-left: auto;
  font-size: 1.2rem;
  line-height: 1.6rem;
  color: ${COLORS.GRAYSCALE_8};
`

export {
  TextWrapper,
  TitleWithIconWrapper,
  NameSpan,
  PageSpan
}
