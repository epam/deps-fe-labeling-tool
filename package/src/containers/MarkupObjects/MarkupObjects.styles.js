
import styled from 'styled-components'
import { COLORS } from '@/theme/theme.default'

const GroupTitle = styled.div`
  margin: 0.5rem 0;
  font-weight: 600;
  font-size: 1.2rem;
  line-height: 1.2rem;
  color: ${COLORS.GRAYSCALE_5};
`

const GapWrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1.6rem;
`

export {
  GapWrapper,
  GroupTitle
}
