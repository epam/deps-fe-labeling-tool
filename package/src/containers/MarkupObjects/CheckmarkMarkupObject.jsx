import React, { useCallback, useMemo } from 'react'
import { PropTypes } from 'prop-types'
import { connect } from 'react-redux'
import { selectLabels } from '@/actions/markup'
import { fieldShape } from '@/models/Field'
import { labelShape } from '@/models/Label'
import { Markup, markupShape } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { MarkupObjectHeader } from './MarkupObjectHeader'

const CheckmarkMarkupObject = ({
  markup,
  label,
  field,
  selectLabels
}) => {
  const objectPage = useMemo(
    () => Markup.getPageByObjectUid(markup, label.uid),
    [label.uid, markup]
  )

  const onHeaderClick = useCallback(() => {
    selectLabels(objectPage, [label])
  }, [label, objectPage, selectLabels])

  return (
    <MarkupObjectHeader
      key={label.uid}
      page={objectPage}
      name={`${field.name} ${label?.index ?? ''}`}
      onHeaderClick={onHeaderClick}
    />
  )
}

CheckmarkMarkupObject.propTypes = {
  markup: markupShape,
  label: labelShape,
  field: fieldShape,
  selectLabels: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  markup: markupSelector(state)
})

const mapDispatchToProps = {
  selectLabels
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(CheckmarkMarkupObject)

export {
  ConnectedComponent as CheckmarkMarkupObject
}
