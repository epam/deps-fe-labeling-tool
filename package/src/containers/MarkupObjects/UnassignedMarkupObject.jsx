import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { selectLabels, selectTables } from '@/actions/markup'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape } from '@/models/Label'
import { Markup, markupShape } from '@/models/Markup'
import { tableShape } from '@/models/Table'
import { markupSelector } from '@/selectors/markup'
import { MarkupObjectHeader } from './MarkupObjectHeader'

const UnassignedMarkupObject = ({
  markupObjects,
  field,
  markup,
  selectLabels,
  selectTables
}) => {
  const getObjectPage = (markup, uid) => Markup.getPageByObjectUid(markup, uid)

  const onHeaderClick = useCallback((object) => {
    const objectPage = getObjectPage(markup, object.uid)

    object.typeName === FieldType.TABLE
      ? selectTables(objectPage, [object])
      : selectLabels(objectPage, [object])
  }, [markup, selectLabels, selectTables])

  return markupObjects.map((markupObject, index) => (
    <MarkupObjectHeader
      key={markupObject.uid}
      page={getObjectPage(markup, markupObject.uid)}
      name={`${field.name} ${index + 1}`}
      onHeaderClick={() => onHeaderClick(markupObject)}
      unassigned
    />
  ))
}

UnassignedMarkupObject.propTypes = {
  markup: markupShape,
  markupObjects: PropTypes.arrayOf(
    PropTypes.oneOfType([
      labelShape,
      tableShape
    ])
  ),
  field: fieldShape,
  selectLabels: PropTypes.func.isRequired,
  selectTables: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  markup: markupSelector(state)
})

const mapDispatchToProps = {
  selectLabels,
  selectTables
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(UnassignedMarkupObject)

export {
  ConnectedComponent as UnassignedMarkupObject
}
