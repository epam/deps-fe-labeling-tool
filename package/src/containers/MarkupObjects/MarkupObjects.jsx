import React, { useCallback, useMemo } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { NothingFound } from '@/components/NothingFound'
import { FieldType } from '@/enums/FieldType'
import { Field, fieldShape } from '@/models/Field'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup, markupShape } from '@/models/Markup'
import { TABLE_TYPE_NAME } from '@/models/Table'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { markupObjectsFilterSelector } from '@/selectors/ui'
import { ListOfMarkupObjects } from './ListOfMarkupObjects'
import {
  GapWrapper,
  GroupTitle
} from './MarkupObjects.styles'
import { PairsMarkupObject } from './PairsMarkupObject'
import { PrimitiveMarkupObjects } from './PrimitiveMarkupObjects'
import { TableMarkupObject } from './TableMarkupObject'
import { UnassignedMarkupObject } from './UnassignedMarkupObject'

const RESOURCE_CHECKMARK_GROUP_NAME = 'CHECKBOXES'
const RESOURCE_LISTS_GROUP_NAME = 'LISTS'
const RESOURCE_TABLE_GROUP_NAME = 'TABLES'
const RESOURCE_PAIR_GROUP_NAME = 'KEY-VALUE PAIRS'
const RESOURCE_STRING_GROUP_NAME = 'STRINGS'
const RESOURCE_ENUM_GROUP_NAME = 'ENUMERATIONS'
const RESOURCE_DATE_GROUP_NAME = 'DATES'
const RESOURCE_UNASSIGNED_GROUP_NAME = 'UNASSIGNED'
const UNASSIGNED_GROUP_CODE = 'unassigned'

const FIELD_TYPE_TO_COMPONENT = {
  [FieldType.CHECKMARK]: (markup, field, index) => <PrimitiveMarkupObjects key={field.code + index} markupObjects={markup} field={field} />,
  [FieldType.LIST]: (markup, field, index) => <ListOfMarkupObjects key={field.code + index} markupObjects={markup} field={field} renderMapper={FIELD_TYPE_TO_COMPONENT} />,
  [FieldType.TABLE]: ([table], field, index) => <TableMarkupObject key={table.uid} table={table} field={field} />,
  [FieldType.PAIR]: (markup, field, index) => <PairsMarkupObject key={field.code + index} markupObjects={markup} field={field} />,
  [FieldType.STRING]: (markup, field, index) => <PrimitiveMarkupObjects key={field.code + index} markupObjects={markup} field={field} />,
  [FieldType.ENUM]: (markup, field, index) => <PrimitiveMarkupObjects key={field.code + index} markupObjects={markup} field={field} />,
  [FieldType.DATE]: (markup, field, index) => <PrimitiveMarkupObjects key={field.code + index} markupObjects={markup} field={field} />
}

const UNASSIGNED_TABLE = {
  NAME: `${LabelType.UNASSIGNED} ${TABLE_TYPE_NAME}`,
  CODE: '_unassigned_tables'
}

const UNASSIGNED_LABEL = {
  NAME: `${LabelType.UNASSIGNED} ${LABEL_TYPE_NAME}`,
  CODE: '_unassigned_labels'
}

const GROUPING_TYPE = {
  VIEW_ALL: 'viewAll',
  FIELD_TYPE: 'fieldType'
}

class SortGroup {
  constructor (code, name, elements) {
    this.code = code
    this.name = name
    this.elements = elements
  }
}

const FIELD_TYPE_TO_GROUP_TITLE = {
  [FieldType.CHECKMARK]: RESOURCE_CHECKMARK_GROUP_NAME,
  [FieldType.ENUM]: RESOURCE_ENUM_GROUP_NAME,
  [FieldType.LIST]: RESOURCE_LISTS_GROUP_NAME,
  [FieldType.TABLE]: RESOURCE_TABLE_GROUP_NAME,
  [FieldType.PAIR]: RESOURCE_PAIR_GROUP_NAME,
  [FieldType.STRING]: RESOURCE_STRING_GROUP_NAME,
  [FieldType.DATE]: RESOURCE_DATE_GROUP_NAME
}

const UNASSIGNED_FIELD_TABLE = new Field(UNASSIGNED_TABLE.CODE, UNASSIGNED_TABLE.NAME, FieldType.TABLE)
const UNASSIGNED_FIELD_LABEL = new Field(UNASSIGNED_LABEL.CODE, UNASSIGNED_LABEL.NAME, FieldType.STRING)

const DEFAULT_UNASSIGNED_COMPONENT = (markup, field) => (
  <UnassignedMarkupObject
    key={field.name}
    markupObjects={markup}
    field={field}
  />
)

const VIEW_GROUP_TO_TABS = {
  [GROUPING_TYPE.VIEW_ALL]: (toRender) => (
    <GapWrapper>
      {
        toRender.map(([config, markupObjects], i) => {
          const renderer = getRenderer(config)
          return renderer(markupObjects, config, i)
        })
      }
    </GapWrapper>
  ),
  [GROUPING_TYPE.FIELD_TYPE]: (toRender) => {
    const groupsToRender = toRender.reduce((acc, [config, markupObjects], i) => {
      const renderer = getRenderer(config)

      if (isUnassigned(config)) {
        const group = acc.find((g) => g.code === UNASSIGNED_GROUP_CODE)
        group
          ? group.elements = [...group.elements, renderer(markupObjects, config)]
          : acc = [
            ...acc,
            new SortGroup(
              UNASSIGNED_GROUP_CODE,
              RESOURCE_UNASSIGNED_GROUP_NAME,
              renderer(markupObjects, config, i)
            )
          ]
        return acc
      }

      const group = acc.find((g) => g.code === config.fieldType)
      group
        ? group.elements = [...group.elements, renderer(markupObjects, config)]
        : acc = [
          ...acc,
          new SortGroup(
            config.fieldType,
            FIELD_TYPE_TO_GROUP_TITLE[config.fieldType],
            [renderer(markupObjects, config, i)]
          )
        ]
      return acc
    }, [])

    return groupsToRender.map((group) => (
      <React.Fragment key={group.code}>
        <GroupTitle>
          {group.name}
        </GroupTitle>
        <GapWrapper>
          {group.elements}
        </GapWrapper>
      </React.Fragment>
    ))
  }
}

const isUnassigned = (field) => (
  field === UNASSIGNED_FIELD_TABLE ||
  field === UNASSIGNED_FIELD_LABEL
)

const getRenderer = (field) => {
  if (isUnassigned(field)) {
    return DEFAULT_UNASSIGNED_COMPONENT
  }

  return FIELD_TYPE_TO_COMPONENT[field.fieldType]
}

const getMarkupToRender = (markup, fields) => {
  const result = []

  fields.forEach((field) => {
    const markupToRender = markup.filter((markupObject) => (
      markupObject.fieldCode === field.code ||
      (
        field.code === UNASSIGNED_LABEL.CODE &&
        markupObject.fieldCode === '' &&
        markupObject.type === LabelType.UNASSIGNED
      ) ||
      (
        field.code === UNASSIGNED_TABLE.CODE &&
        markupObject.fieldCode === '' &&
        markupObject.typeName === TABLE_TYPE_NAME
      )
    ))

    if (markupToRender.length) {
      result.push([
        field,
        markupToRender
      ])
    }
  })

  return result
}

const MarkupObjects = ({
  fields,
  markupObjectsFilter,
  fieldsGrouping,
  markup,
  setFilter
}) => {
  const filteredFields = useMemo(
    () => {
      const allFields = [...fields, UNASSIGNED_FIELD_LABEL, UNASSIGNED_FIELD_TABLE]
      return allFields.filter((f) => f.name.toLowerCase().includes(markupObjectsFilter.toLowerCase()))
    },
    [fields, markupObjectsFilter]
  )

  const resetFilter = useCallback(() => {
    setFilter('')
  }, [setFilter])

  if (!markup) {
    return null
  }

  if (!filteredFields.length) {
    return <NothingFound resetFilter={resetFilter} />
  }

  const allMarkupObjects = Markup.getAllObjects(markup)
  const toRender = getMarkupToRender(allMarkupObjects, filteredFields)
  return VIEW_GROUP_TO_TABS[fieldsGrouping](toRender)
}

MarkupObjects.propTypes = {
  markup: markupShape,
  fields: PropTypes.arrayOf(fieldShape),
  markupObjectsFilter: PropTypes.string.isRequired,
  fieldsGrouping: PropTypes.string,
  setFilter: PropTypes.func.isRequired
}

const mapStateToProps = (state) => ({
  markup: markupSelector(state),
  fields: fieldsSelector(state),
  markupObjectsFilter: markupObjectsFilterSelector(state)
})

const ConnectedComponent = connect(mapStateToProps)(MarkupObjects)

export {
  ConnectedComponent as MarkupObjects
}
