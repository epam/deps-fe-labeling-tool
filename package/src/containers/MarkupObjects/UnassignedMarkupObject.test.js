import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockUuid } from '@/mocks/mockUuid'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import React from 'react'
import { shallow } from 'enzyme'
import { MarkupObjectHeader } from '@/containers/MarkupObjects/MarkupObjectHeader'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { LabelType, LABEL_TYPE_NAME, Label } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { Table } from '@/models/Table'
import { UnassignedMarkupObject } from './UnassignedMarkupObject'

jest.mock('uuid', () => mockUuid)
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('@/models/Markup', () => ({
  ...jest.requireActual('@/models/Markup'),
  Markup: {
    getPageByObjectUid: jest.fn(() => '1')
  }
}))
jest.mock('./MarkupObjectHeader', () => mockComponent('MarkupObjectHeader'))

const { WrappedComponent, mapStateToProps } = UnassignedMarkupObject

describe('Component: UnassignedMarkupObject', () => {
  describe('mapStateToProps', () => {
    it('should pass state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      const expectedProps = {
        markup: mockMarkupSelectors.markupSelector()
      }

      expect(props).toEqual(expectedProps)
    })
  })

  describe('component', () => {
    let defaultProps, wrapper

    const mockObjectPage = Markup.getPageByObjectUid()

    beforeEach(() => {
      defaultProps = {
        field: new Field('_testCode', LABEL_TYPE_NAME, FieldType.STRING),
        markupObjects: [
          new Label(
            1,
            2,
            3,
            4,
            '_testCode',
            undefined,
            LabelType.UNASSIGNED,
            ''
          )
        ],
        markup: mockMarkupSelectors.markupSelector(),
        selectTables: jest.fn(),
        selectLabels: jest.fn()
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call selectLabels with correct args in case of calling  MarkupObjectHeader prop onHeaderClick', () => {
      wrapper.find(MarkupObjectHeader).props().onHeaderClick()
      expect(defaultProps.selectLabels).nthCalledWith(1, mockObjectPage, defaultProps.markupObjects)
    })

    it('should call selectTables with correct args in case of calling  MarkupObjectHeader prop onHeaderClick', () => {
      defaultProps.markupObjects = [new Table(
        [10, 20],
        [10, 20]
      )]
      wrapper.setProps(defaultProps)

      wrapper.find(MarkupObjectHeader).props().onHeaderClick()
      expect(defaultProps.selectTables).nthCalledWith(1, mockObjectPage, defaultProps.markupObjects)
    })
  })
})
