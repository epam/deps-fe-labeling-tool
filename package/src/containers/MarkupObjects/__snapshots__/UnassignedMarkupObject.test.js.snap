// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: UnassignedMarkupObject component should render correct layout 1`] = `
<MarkupObjectHeader
  key="6"
  name="label 1"
  onHeaderClick={[Function]}
  page="1"
  unassigned={true}
/>
`;
